### PR TITLE
Implement context_filter_keypress to always commit the pressed key

### PR DESCRIPTION
**Pull request by awhiemstra**
Apparently this is needed now, otherwise we get no text at all.

### CODE REVIEW

**Comment by ebassi** [comment]
Quick drive-by comment: this should not be necessary, or, at the very least, it would indicate something has broken inside GTK3 which shouldn't have been broken (GTK3 is API/feature frozen, so any change in existing working, idiomatic code is considered a breakage).

The default implementation of `GtkIMContextClass.filter_keypress` returns "false", meaning that the input method has not consumed the key event, and thus the toolkit is responsible for it. The `commit()` API is meant to be used to inject text generated by the input method into the editable widget.

**Comment by garnacho** [comment]
GTK/IM developer here. This does indeed sound off as @ebassi said, and the stock GTK3 GtkIMContext implementation should be able to handle UTF8 text committed directly from the IM/compositor. I would be interested to see `WAYLAND_DEBUG=1` logs of an app reproducing this issue.

**Comment by bertogg** [comment]
I don't have previous IM experience, but after having a quick look at this (using GtkEntry in particular) my impressions are (correct me if they're wrong):

- If `gtk_im_context_filter_keypress()` returns FALSE [here](https://gitlab.gnome.org/GNOME/gtk/-/blob/3.24.39/gtk/gtkentry.c?ref_type=tags#L4898) then GTK only handles the key bindings, not normal text input.
- `GtkIMContextXIM` does implement `_filter_keypress` [here](https://gitlab.gnome.org/GNOME/gtk/-/blob/3.24.39/modules/input/gtkimcontextxim.c#L637).
- Other modules such as `GtkIMContextWayland` or `GtkIMContextBroadway` don't implement `_filter_keypress()` but derive from `GtkIMContextSimple`, which [does](https://gitlab.gnome.org/GNOME/gtk/-/blob/3.24.39/gtk/gtkimcontextsimple.c#L685).
- If `GTK_IM_MODULE` is unset then a default module is picked by [_gtk_im_module_get_default_context_id()](https://gitlab.gnome.org/GNOME/gtk/-/blob/3.24.39/gtk/gtkimmodule.c#L829) (`ibus`, in my tests).
- Because of the last point I don't know if the stock `GtkIMContext` is ever used in practice ?

About `WAYLAND_DEBUG`, for @garnacho: this is happening with X11.

**Comment by bertogg** [comment]
FWIW my first test when I was debugging this was to simply make `SteamIMContext` derive from `GtkIMContextSimple` instead of `GtkIMContext`, and that seems to work, at least for a basic use case with a text entry.

**Comment by bertogg** [comment]
Here's a basic test case to reproduce the problem on a regular desktop: [test-im-context.c](https://github.com/valve-project/steam-qt-keyboard-plugin/files/13953016/test-im-context.c.txt)

Build and install the plugin and run Zenity like this:
```
$ GTK_IM_MODULE=Test zenity --entry --entry-text=foobar
```
You should be able to enter text normally, but if you change `g_type_module_register_type()` and derive from `GTK_TYPE_IM_CONTEXT` instead it'll stop working.

**Comment by ebassi** [comment]
Mystery solved:

```c
static gboolean
test_im_context_filter_keypress (GtkIMContext *context,
                                 GdkEventKey  *event)
{
    if (event->type == GDK_KEY_RELEASE) {
        gunichar ch = gdk_keyval_to_unicode (event->keyval);
        if (ch != 0 && !g_unichar_iscntrl (ch)) {
            char buf[8] = { 0, };
            g_unichar_to_utf8 (ch, buf);
            g_signal_emit_by_name (context, "commit", buf);
            return TRUE;
        }
    }
    return FALSE;
}
```

When inheriting from IMContext, we need to override `GtkIMContextClass.filter_keypress` and commit the Unicode characters ourselves.

**Comment by ebassi** [comment]
I'd recommend not inheriting from `GtkIMContextSimple`, as it adds a bunch of behaviours we might not want, like Compose key support and Unicode codepoint input.

File: gtk/steam-imcontext-gtk.c [hunk 1]
```diff
@@ -129,3 +125,34 @@ static void steam_im_context_focus_in(GtkIMContext *context)
         g_warning("Error opening Steam keyboard: %s", error->message);
     }
 }
+
+static gboolean steam_im_context_filter_keypress(GtkIMContext *context, GdkEventKey *event)
+{
+    SteamIMContext *im_context = STEAM_IM_CONTEXT(context);
+
+    if (event->type == GDK_KEY_RELEASE) {
+        if (event->keyval == GDK_KEY_BackSpace) {
```
Comment (ebassi) [inline gtk/steam-imcontext-gtk.c:134]: You don't need to handle backspace/delete/and arrow keys: those are managed by GTK itself when returning FALSE from `filter_keypress`.

The only thing that's needed is the emission of the "commit" signal with the Unicode glyph converted into a string.

File: gtk/steam-imcontext-gtk.c [hunk 2]
```diff
@@ -129,3 +125,34 @@ static void steam_im_context_focus_in(GtkIMContext *context)
...
+	    return TRUE;
+	}
+
+	gchar keyval_utf8[7];
+	gint length = g_unichar_to_utf8(gdk_keyval_to_unicode(event->keyval), keyval_utf8);
+	keyval_utf8[length] = '\0';
+
+        g_signal_emit_by_name(im_context, "commit", keyval_utf8);
+
+	return TRUE;
```
Comment (ebassi) [inline gtk/steam-imcontext-gtk.c:153]: Need to check for a valid/non-control Unicode character before committing it:

```suggestion
        gunichar ch = gdk_keyval_to_unicode(event->keyval);
        if (ch != 0 && !g_unichar_iscntl(ch)) {
            gchar keyval_utf8[8] = { 0, };
	    g_unichar_to_utf8(ch, keyval_utf8);
	    g_signal_emit_by_name(im_context, "commit", keyval_utf8);
	    return TRUE;
	}
```

**Comment by bertogg** [comment]
I tested the updated branch with commit 872b047a74545c0de4ca85989339a1d1410c44af and it seems to work fine here.

**Pull request merged**